File: crm/src/main/kotlin/com/example/crm/entities/Customer.kt
package com.example.crm.entities

import jakarta.persistence.Entity
import jakarta.persistence.GeneratedValue
import jakarta.persistence.Id
import jakarta.persistence.OneToMany
import jakarta.persistence.OneToOne

@Entity
class Customer {
    @Id
    @GeneratedValue
    val customerId: Long = 0

    @OneToMany(mappedBy = "currentCustomer")
    var jobOffers: MutableSet<JobOffer> = mutableSetOf()

    @OneToMany(mappedBy = "oldCustomer")
    var replacementHistory : MutableSet<JobOffer> = mutableSetOf()

    @OneToOne
    var contact: Contact? = null//tmp :Contact 1to1

    lateinit var notes : String


    fun removeJobOffer(jo: JobOffer){
        jo.currentCustomer = null
        jobOffers.remove(jo)
    }

    fun completeJobOffer(jo: JobOffer){
        jobOffers.remove(jo)
        jo.currentCustomer = null
        jo.oldCustomer = this
        replacementHistory.add(jo)
    }

    fun addContact(c : Contact){
        c.customer = this
        contact = c
    }
}

File: frontend/src/api/crm/entities/Customer.ts
import {JobOffer} from "./JobOffer";

interface CustomerRawData {
    customerId: number | null
    jobOffers: JobOffer[] | null
    replacementHistory: JobOffer[] | null
    notes: string[] | null
}

export class Customer implements CustomerRawData {
    customerId: number | null
    jobOffers: JobOffer[] | null
    replacementHistory: JobOffer[] | null
    notes: string[] | null

    constructor(customerId: number | null, jobOffers: JobOffer[] | null, replacementHistory: JobOffer[]| null, notes: string[] | null) {
        this.customerId = customerId
        this.jobOffers = jobOffers
        this.replacementHistory = replacementHistory
        this.notes = notes
    }

    static fromJsonObject(obj: CustomerRawData): Customer | null {
        try {
            return new Customer(obj.customerId, obj.jobOffers, obj.replacementHistory, obj.notes)
        } catch (e) {
            console.error(e)
            return null
        }
    }
}
